File: sdk/python/openbrain_sdk/openbrain_sdk/mcp.py
from __future__ import annotations

import json
import queue
import subprocess
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Optional, TypeAlias

from .error import OpenBrainError

RpcRequest: TypeAlias = dict[str, Any]
RpcResponse: TypeAlias = dict[str, Any]


@dataclass
class OpenBrainMcpClient:
    proc: subprocess.Popen
    timeout_sec: float = 10.0
    _responses: "queue.Queue[tuple[Optional[int], RpcResponse]]" = field(default_factory=queue.Queue)
    _next_id: int = 1
    _closed: bool = False

    @classmethod
    def spawn_openbrain_mcp(
        cls,
        openbrain_path: str = "openbrain",
        args: Optional[list[str]] = None,
        env: Optional[dict[str, str]] = None,
        cwd: Optional[str] = None,
        startup_timeout_sec: float = 5.0,
    ) -> "OpenBrainMcpClient":
        proc = subprocess.Popen(
            [openbrain_path, *(args or [])],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            env=env,
            cwd=cwd,
        )
        if proc.stdin is None or proc.stdout is None:
            raise OpenBrainError("MCP_ERROR", "failed to create stdio pipes")

        client = cls(proc=proc)
        threading.Thread(target=cls._reader_loop, args=(proc, client._responses), daemon=True).start()
        start = time.time()
        while proc.poll() is None and time.time() - start < startup_timeout_sec:
            time.sleep(0.05)
        if proc.poll() is not None:
            raise OpenBrainError("MCP_ERROR", "MCP process exited during startup")
        return client

    @staticmethod
    def _reader_loop(proc: subprocess.Popen, q: "queue.Queue[tuple[Optional[int], RpcResponse]]") -> None:
        if proc.stdout is None:
            return
        for line in proc.stdout:
            line = line.strip()
            if not line:
                continue
            try:
                payload = json.loads(line)
            except json.JSONDecodeError:
                continue
            msg_id = payload.get("id")
            q.put((msg_id, payload))

    def call_tool(self, tool: str, args: dict[str, Any], timeout_sec: Optional[float] = None) -> Any:
        if self._closed:
            raise OpenBrainError("MCP_CLOSED", "MCP client is closed")

        if self.proc.stdin is None:
            raise OpenBrainError("MCP_ERROR", "stdin is unavailable")

        msg_id = self._next_id
        self._next_id += 1
        req: RpcRequest = {
            "jsonrpc": "2.0",
            "id": msg_id,
            "method": "tools/call",
            "params": {"name": tool, "arguments": args},
        }
        self.proc.stdin.write(json.dumps(req) + "\n")
        self.proc.stdin.flush()

        target_timeout = timeout_sec if timeout_sec is not None else self.timeout_sec
        try:
            while True:
                response_id, payload = self._responses.get(timeout=target_timeout)
                if response_id == msg_id:
                    break
        except queue.Empty as err:
            raise OpenBrainError("TIMEOUT", "MCP call timed out") from err

        if not isinstance(payload, dict):
            raise OpenBrainError("MCP_ERROR", "invalid MCP payload")
        if "error" in payload:
            raise OpenBrainError("MCP_ERROR", payload["error"].get("message", "unknown"), details=payload["error"])
        result = payload.get("result")
        if not isinstance(result, dict):
            raise OpenBrainError("MCP_ERROR", "missing result payload")
        if result.get("ok") is False:
            error = result.get("error")
            if not isinstance(error, dict):
                raise OpenBrainError("OB_INTERNAL", "invalid error payload", details=result)
            raise OpenBrainError(error.get("code", "OB_INTERNAL"), error.get("message", "error"), details=error.get("details"))

        return result

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self.proc.stdin:
            self.proc.stdin.close()
        self.proc.terminate()
        try:
            self.proc.wait(timeout=2.0)
        except subprocess.TimeoutExpired:
            self.proc.kill()

File: sdk/typescript/openbrain-sdk/src/mcp.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { createInterface } from "node:readline";
import { once } from "node:events";
import { OpenBrainError } from "./error.js";

type JsonRpcId = number;
type PendingResolver = (payload: unknown) => void;
type PendingRejector = (error: OpenBrainError) => void;

interface RpcRequest {
  jsonrpc: "2.0";
  id: JsonRpcId;
  method: string;
  params?: unknown;
}

interface RpcResponse {
  jsonrpc: "2.0";
  id: JsonRpcId | null;
  result?: unknown;
  error?: {
    code: number;
    message: string;
    data?: unknown;
  };
}

interface OpenBrainMcpClientOptions {
  openbrainPath?: string;
  args?: string[];
  env?: NodeJS.ProcessEnv;
  cwd?: string;
  startupTimeoutMs?: number;
}

interface PendingCall {
  resolve: PendingResolver;
  reject: PendingRejector;
}

const DEFAULT_COMMAND = "openbrain";

export class OpenBrainMcpClient {
  private readonly proc: ChildProcessWithoutNullStreams;
  private readonly openbrainPath: string;
  private readonly startupTimeoutMs: number;
  private readonly pending = new Map<JsonRpcId, PendingCall>();
  private nextId = 1;
  private closed = false;

  constructor(proc: ChildProcessWithoutNullStreams, openbrainPath: string, startupTimeoutMs: number) {
    this.proc = proc;
    this.openbrainPath = openbrainPath;
    this.startupTimeoutMs = startupTimeoutMs;
  }

  static async spawnOpenBrainMcp(
    options: OpenBrainMcpClientOptions = {}
  ): Promise<OpenBrainMcpClient> {
    const openbrainPath = options.openbrainPath ?? DEFAULT_COMMAND;
    const proc = spawn(openbrainPath, options.args ?? [], {
      stdio: ["pipe", "pipe", "pipe"],
      env: options.env ?? process.env,
      cwd: options.cwd,
      windowsHide: true,
    });

    if (!proc.stdin || !proc.stdout) {
      throw new Error(`failed to start MCP child process: ${openbrainPath}`);
    }

    const client = new OpenBrainMcpClient(
      proc,
      openbrainPath,
      options.startupTimeoutMs ?? 5000
    );
    client.bindResponseReader();
    await client.waitForTransportReady();
    return client;
  }

  private bindResponseReader() {
    const rl = createInterface({ input: this.proc.stdout });
    rl.on("line", (line) => {
      let response: RpcResponse;
      try {
        response = JSON.parse(line) as RpcResponse;
      } catch (_err) {
        return;
      }

      if (response.id == null) {
        return;
      }

      const pending = this.pending.get(response.id);
      if (!pending) {
        return;
      }
      this.pending.delete(response.id);

      if (response.error) {
        pending.reject(
          new OpenBrainError("MCP_ERROR", response.error.message, {
            details: { code: response.error.code, data: response.error.data },
          })
        );
        return;
      }

      if (
        typeof response.result === "object" &&
        response.result !== null &&
        "ok" in (response.result as Record<string, unknown>) === false
      ) {
        pending.resolve(response.result);
        return;
      }

      const result = response.result as {
        ok?: boolean;
        error?: { code: string; message: string; details?: unknown };
      };

      if (result?.ok === false && result?.error) {
        pending.reject(
          new OpenBrainError(result.error.code, result.error.message, {
            details: result.error.details,
          })
        );
        return;
      }

      pending.resolve(response.result);
    });

    this.proc.on("exit", () => {
      for (const [, pending] of this.pending) {
        pending.reject(
          new OpenBrainError("MCP_PROCESS_EXITED", "MCP process exited before response")
        );
      }
      this.pending.clear();
    });

    this.proc.on("error", (err) => {
      for (const [, pending] of this.pending) {
        pending.reject(new OpenBrainError("MCP_PROCESS_ERROR", err.message));
      }
      this.pending.clear();
    });
  }

  private async waitForTransportReady() {
    return Promise.race([
      new Promise<void>((resolve) => {
        this.proc.once("spawn", () => resolve());
      }),
      new Promise<never>((_, reject) =>
        setTimeout(() => reject(new OpenBrainError("TIMEOUT", "MCP process failed to spawn")), this.startupTimeoutMs)
      ),
    ]);
  }

  private sendRequest(method: string, params: Record<string, unknown>): Promise<unknown> {
    if (this.closed) {
      return Promise.reject(new OpenBrainError("MCP_CLOSED", "MCP session is closed"));
    }
    const id = this.nextId++;
    const request: RpcRequest = {
      jsonrpc: "2.0",
      id,
      method,
      params,
    };

    return new Promise<unknown>((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      this.proc.stdin.write(`${JSON.stringify(request)}\n`, "utf8", (error) => {
        if (error) {
          this.pending.delete(id);
          reject(new OpenBrainError("MCP_WRITE_ERROR", error.message));
        }
      });
    });
  }

  async callTool<TToolName extends keyof import("./types.js").OpenBrainToolResponseMap>(
    toolName: TToolName,
    args: import("./types.js").OpenBrainToolRequestMap[TToolName],
    options?: { timeoutMs?: number }
  ): Promise<import("./types.js").OpenBrainToolResponseMap[TToolName]> {
    const result = await Promise.race([
      this.sendRequest("tools/call", {
        name: toolName,
        arguments: args,
      }),
      new Promise<never>((_, reject) =>
        setTimeout(
          () => reject(new OpenBrainError("TIMEOUT", "MCP tool call timeout")),
          options?.timeoutMs ?? 10_000
        )
      ),
    ]);

    return result as import("./types.js").OpenBrainToolResponseMap[TToolName];
  }

  async close() {
    if (this.closed) return;
    this.closed = true;
    this.proc.stdin.end();
    this.proc.kill();
    await once(this.proc, "exit").catch(() => undefined);
  }

}
